src/webgpu/shader/execution/expression/case_cache.ts

```diff
@@ -0,0 +1,202 @@
+import { Cacheable, dataCache } from '../../../../common/framework/data_cache.js';
+import { SerializedComparator, deserializeComparator } from '../../../util/compare.js';
+import {
+  Scalar,
+  Vector,
+  serializeValue,
+  SerializedValue,
+  deserializeValue,
+} from '../../../util/conversion.js';
+import {
+  deserializeF32Interval,
+  F32Interval,
+  SerializedF32Interval,
+  serializeF32Interval,
+} from '../../../util/f32_interval.js';
+
+import { Case, CaseList, Expectation } from './expression.js';
+
+/**
+ * SerializedExpectationValue holds the serialized form of an Expectation when
+ * the Expectation is a Value
+ * This form can be safely encoded to JSON.
+ */
+type SerializedExpectationValue = {
+  kind: 'value';
+  value: SerializedValue;
+};
+
+/**
+ * SerializedExpectationValue holds the serialized form of an Expectation when
+ * the Expectation is an Interval
+ * This form can be safely encoded to JSON.
+ */
+type SerializedExpectationInterval = {
+  kind: 'interval';
+  value: SerializedF32Interval;
+};
+
+/**
+ * SerializedExpectationValue holds the serialized form of an Expectation when
+ * the Expectation is a list of Intervals
+ * This form can be safely encoded to JSON.
+ */
+type SerializedExpectationIntervals = {
+  kind: 'intervals';
+  value: SerializedF32Interval[];
+};
+
+/**
+ * SerializedExpectationValue holds the serialized form of an Expectation when
+ * the Expectation is a Comparator
+ * This form can be safely encoded to JSON.
+ */
+type SerializedExpectationComparator = {
+  kind: 'comparator';
+  value: SerializedComparator;
+};
```

Comment:
optional idea:
I'm not sure if it would make any meaningful performance difference, but the JSON output could be much shorter by using something like:
`{ comparator: SerializedComparator }`
or a tuple type:
`['comparator', SerializedComparator]`

Similarly e.g. `{begin, end}` could be `[begin, end]`